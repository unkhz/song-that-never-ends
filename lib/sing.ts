function removeIndentation(str: string) {
  return str.replace(/\n\s+/g, "\n");
}

async function wait(delay: number) {
  const variance = Math.round(delay * 0.33);
  const naturalDelay =
    delay + (Math.round(variance * 2 * Math.random()) - variance);
  await new Promise((resolve) => setTimeout(resolve, naturalDelay));
}

async function type(line: string, delay: number) {
  for (const char of line) {
    await wait(delay);
    process.stdout.write(char);
  }
}

async function say(line: string, voice: string) {
  const proc = Bun.spawn(["say", `-v${voice}`, line]);
  await proc.exited;
}

async function sayAndType(line: string, voice: string, delay: number) {
  await Promise.all([type(`${line}\n`, delay), say(line, voice)]);
}

export async function sing(
  voice: string,
  verse1: string,
  verse2: string,
  verse3: string,
  chorus: string,
  finalChorus: string
) {
  const parts = [verse1, chorus, verse2, chorus, verse3, finalChorus];
  for (const part of parts.map(removeIndentation)) {
    for (const line of part.split("\n")) {
      const trimmedLine = line.trim();
      if (trimmedLine) {
        await sayAndType(line, voice, 66);
      }
    }
    await type(`\n`, 11);
  }
  await type(`-------\n`, 66);
}
