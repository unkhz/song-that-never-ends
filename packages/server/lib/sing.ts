import { typeLine, wait } from 'tools'

async function drain(it: AsyncIterableIterator<unknown>) {
  for await (const _ of it) {
    // just drain it, we don't care about the values
  }
}

async function* singPart(part: string, typingDelay: number) {
  for (const line of part.split('\n')) {
    yield `say:${line}`
    const trimmedLine = line.trim()
    if (trimmedLine) {
      for await (const char of typeLine(line, typingDelay)) {
        yield char
      }
    }
  }
  await wait(typingDelay * 10)
  yield '\n'
  await wait(typingDelay * 10)
}

// After decillion iterations, start over again. What else can you do?
// According to Wikipedia (https://en.wikipedia.org/wiki/Names_of_large_numbers),
// numbers after decillion are not widely standardized, so it's perhaps the
// best time to start over.
const loopLength = 1_000_000_000_000_000_000_000_000_000_000_000_000_000_000n

export async function* sing(
  song: (iteration: bigint) => Promise<string[]>,
  typingDelay: number
) {
  let iteration = 0n
  let nextParts: Promise<string[]> = song(iteration)
  while (true) {
    const currentParts = await nextParts
    iteration++
    nextParts = song(iteration)

    for (const part of currentParts) {
      for await (const char of singPart(part, typingDelay)) {
        yield char
      }
    }

    if (iteration === loopLength) {
      iteration = 0n
    }
  }
}
