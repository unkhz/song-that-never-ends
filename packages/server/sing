#!/bin/bash

clear

voice=${1:-"Daniel"}

# Set terminal title
echo -en "\033]0;Song that never ends\007"

bun index.ts "$voice"