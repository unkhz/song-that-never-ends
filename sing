#!/bin/bash

clear

# Set terminal title
echo -en "\033]0;Song that never ends\007"

bun index.ts Daniel