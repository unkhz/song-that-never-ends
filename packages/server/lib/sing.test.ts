import { SongPart } from 'packages/song/index.js'
import { sing } from './sing.js'

describe('sing', () => {
  it('should sing', async () => {
    const song = async (i: bigint): Promise<SongPart[]> => [
      { type: 'chorus', index: 0, content: `iteration ${i}\nsecond line` },
      { type: 'verse', index: 1, content: 'la la la' },
      { type: 'verse', index: 2, content: 'laa laa laa' },
    ]
    const chars = []
    let i = 0
    for await (const char of sing(song, 0)) {
      chars.push(char)
      if (i++ > 74) break
    }
    expect(chars).toMatchSnapshot()
  })
})
