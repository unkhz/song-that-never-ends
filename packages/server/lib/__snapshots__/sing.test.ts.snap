// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sing should sing 1`] = `
[
  {
    "content": 1n,
    "type": "iteration",
  },
  {
    "content": "chorus",
    "type": "part",
  },
  {
    "content": "iteration 0",
    "type": "line",
  },
  {
    "content": "i",
    "type": "char",
  },
  {
    "content": "t",
    "type": "char",
  },
  {
    "content": "e",
    "type": "char",
  },
  {
    "content": "r",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "t",
    "type": "char",
  },
  {
    "content": "i",
    "type": "char",
  },
  {
    "content": "o",
    "type": "char",
  },
  {
    "content": "n",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "0",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "second line",
    "type": "line",
  },
  {
    "content": "s",
    "type": "char",
  },
  {
    "content": "e",
    "type": "char",
  },
  {
    "content": "c",
    "type": "char",
  },
  {
    "content": "o",
    "type": "char",
  },
  {
    "content": "n",
    "type": "char",
  },
  {
    "content": "d",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "i",
    "type": "char",
  },
  {
    "content": "n",
    "type": "char",
  },
  {
    "content": "e",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "verse",
    "type": "part",
  },
  {
    "content": "la la la",
    "type": "line",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "verse",
    "type": "part",
  },
  {
    "content": "laa laa laa",
    "type": "line",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "l",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": 2n,
    "type": "iteration",
  },
  {
    "content": "chorus",
    "type": "part",
  },
  {
    "content": "iteration 1",
    "type": "line",
  },
  {
    "content": "i",
    "type": "char",
  },
  {
    "content": "t",
    "type": "char",
  },
  {
    "content": "e",
    "type": "char",
  },
  {
    "content": "r",
    "type": "char",
  },
  {
    "content": "a",
    "type": "char",
  },
  {
    "content": "t",
    "type": "char",
  },
  {
    "content": "i",
    "type": "char",
  },
  {
    "content": "o",
    "type": "char",
  },
  {
    "content": "n",
    "type": "char",
  },
  {
    "content": " ",
    "type": "char",
  },
  {
    "content": "1",
    "type": "char",
  },
  {
    "content": "
",
    "type": "char",
  },
  {
    "content": "second line",
    "type": "line",
  },
  {
    "content": "s",
    "type": "char",
  },
  {
    "content": "e",
    "type": "char",
  },
  {
    "content": "c",
    "type": "char",
  },
  {
    "content": "o",
    "type": "char",
  },
]
`;
